components/raftstore/src/store/peer_storage.rs

```diff
@@ -767,12 +821,188 @@ where
         Ok(())
     }
 
+    pub fn clean_async_fetch_res(&mut self, low: u64) {
+        self.async_fetch_results.borrow_mut().remove(&low);
+    }
+
+    pub fn set_async_fetch_res(&mut self, low: u64, res: Option<RaftlogFetchResult>) {
+        let in_fetching = match self.async_fetch_results.borrow().get(&low) {
+            Some(RaftlogFetchState::Fetched(_)) => false,
+            Some(RaftlogFetchState::Fetching) => true,
+            None => false,
+        };
+
+        match res {
+            Some(res) => {
+                if let Some(RaftlogFetchState::Fetched(prev)) = self
+                    .async_fetch_results
+                    .borrow_mut()
+                    .insert(low, RaftlogFetchState::Fetched(res))
+                {
+                    warn!(
+                        "unconsumed async fetch res";
+                        "region_id" => self.region.get_id(),
+                        "peer_id" => self.peer_id,
+                        "res" => ?prev,
+                        "low" => low,
+                    );
+                }
+            }
+            None => {
+                if !in_fetching {
+                    let prev = self.async_fetch_results.borrow_mut().remove(&low);
+                    if prev.is_some() {
+                        self.raftlog_fetch_stats.fetch_invalid.update(|m| m + 1);
+                    }
+                }
+            }
+        }
+    }
+
+    fn async_fetch(
+        &self,
+        region_id: u64,
+        low: u64,
+        mut high: u64,
+        max_size: u64,
+        context: GetEntriesContext,
+        buf: &mut Vec<Entry>,
+    ) -> raft::Result<usize> {
+        if let Some(RaftlogFetchState::Fetching) = self.async_fetch_results.borrow().get(&low) {
+            // already an async fetch in flight
+            return Err(raft::Error::Store(
+                raft::StorageError::LogTemporarilyUnavailable,
+            ));
+        }
+
+        let tried_cnt = if let Some(RaftlogFetchState::Fetched(res)) =
+            self.async_fetch_results.borrow_mut().remove(&low)
+        {
+            assert_eq!(res.low, low);
+            let mut ents = res.ents?;
```

Comment:
If `ent` is an error then `raft-rs` will panic, is it expected?

Reply:
yes, the error would be `ErrCompacted` or `ErrUnavailable`. They are possible returned errors that sync API may return.